x/ibc-hooks/contracts/crosschain-swaps/src/execute.rs

```diff
@@ -0,0 +1,271 @@
+        // The response data
+
+        // Add the response data since it won't be added in the reply.
+        let amount = swap_response.amount;
+        let denom = swap_response.token_out_denom;
+        let channel_id = swap_msg_state.forward_to.channel;
+        let to_address = swap_msg_state.forward_to.receiver;
+        let data = CrosschainSwapResponse {
+            msg: format!("Sent {amount}{denom} to {channel_id}/{to_address}"),
+        };
+
+        return Ok(response
+            .set_data(to_binary(&data)?)
+            .add_message(ibc_transfer));
+    }
+
+    // Check that there isn't anything stored in FORWARD_REPLY_STATES. If there
+    // is, it means that the contract is already waiting for a reply and should
+    // not override the stored state. This should never happen here, but adding
+    // the check for safety. If this happens there is likely a malicious attempt
+    // modify the contract's state before it has replied.
+    if FORWARD_REPLY_STATES.may_load(deps.storage)?.is_some() {
+        return Err(ContractError::ContractLocked {
+            msg: "Already waiting for a reply".to_string(),
+        });
+    }
+    // Store the ibc send information and the user's failed delivery preference
+    // so that it can be handled by the response
+    FORWARD_REPLY_STATES.save(
+        deps.storage,
+        &ForwardMsgReplyState {
+            channel_id: swap_msg_state.forward_to.channel,
+            to_address: swap_msg_state.forward_to.receiver.into(),
+            amount: swap_response.amount.into(),
+            denom: swap_response.token_out_denom,
+            failed_delivery: swap_msg_state.forward_to.failed_delivery,
+        },
+    )?;
+
+    Ok(response.add_submessage(SubMsg::reply_on_success(ibc_transfer, FORWARD_REPLY_ID)))
+}
+
+use ::prost::Message; // Proveides ::decode() for MsgTransferResponse
+
+pub fn handle_forward_reply(deps: DepsMut, msg: Reply) -> Result<Response, ContractError> {
+    // Parse the result from the underlying chain call (IBC send)
+    let SubMsgResult::Ok(SubMsgResponse { data: Some(b), .. }) = msg.result else {
+        return Err(ContractError::FailedIBCTransfer { msg: format!("failed reply: {:?}", msg.result) })
+    };
+
+    // The response contains the packet sequence. This is needed to be able to
+    // ensure that, if there is a delivery failure, the packet that failed is
+    // the same one that we stored recovery information for
+    let response =
+        MsgTransferResponse::decode(&b[..]).map_err(|_e| ContractError::FailedIBCTransfer {
+            msg: format!("could not decode response: {b}"),
+        })?;
+
+    // Similar consideration as the warning above. Is it safe for this to be an Item?
+    let ForwardMsgReplyState {
+        channel_id,
+        to_address,
+        amount,
+        denom,
+        failed_delivery,
+    } = FORWARD_REPLY_STATES.load(deps.storage)?;
+    FORWARD_REPLY_STATES.remove(deps.storage);
```

Comment:
What is the concern with being an item? Is there some mutable data in the struct?

Reply:
That's a dangling comment. I've removed it.

Above we're doing
```
    if FORWARD_REPLY_STATES.may_load(deps.storage)?.is_some() {
        return Err(ContractError::ContractLocked {
            msg: "Already waiting for a reply".to_string(),
        });
    }
```

to ensure that there's no "reentrency" (if a submessage would call this contract again). That is not strictly necessary, as the only submessages this contract sends are to swaprouter and ibc, which don't do that. I wasn't checking this before and there used to be a few comments discussing why it wasn't a problem but questioning if it was ok to leave it that way or to check defensively. 

After discussing it with Boss, we concluded it was better to be defensive, so now we check this and that comment isn't needed anymore. 

For background, the issue with it being an Item is that that's basically a singleton, so if we enter the contract again, then the value would be overriden. For contracts that expect this behaviour, we could use a Stack instead (felt silly to use a stack that I know is always gonna have depth 1)